### PR TITLE
[Merged by Bors] - Self rate limiting dev flag

### PR DESCRIPTION
**Pull request by divagant-martian**
## Issue Addressed
Adds self rate limiting options, mainly with the idea to comply with peer's rate limits in small testnets

## Proposed Changes
Add a hidden flag `self-limiter` this can take no value, or customs values to configure quotas per protocol

### How to use
`--self-limiter` will turn on the self rate limiter applying the same params we apply to inbound requests (requests from other peers)
`--self-limiter "beacon_blocks_by_range:64/1"` will turn on the self rate limiter for ALL protocols, but change the quota for bbrange to 64 requested blocks per 1 second.
`--self-limiter "beacon_blocks_by_range:64/1;ping:1/10"` same as previous one, changing the quota for ping as well.

### Caveats
- The rate limiter is either on or off for all protocols. I added the custom values to be able to change the quotas per protocol so that some protocols can be given extremely loose or tight quotas. I think this should satisfy every need even if we can't technically turn off rate limits per protocol.
- This reuses the rate limiter struct for the inbound requests so there is this ugly part of the code in which we need to deal with the inbound only protocols (light client stuff) if this becomes too ugly as we add lc protocols, we might want to split the rate limiters. I've checked this and looks doable with const generics to avoid so much code duplication

### Knowing if this is on
```
Feb 06 21:12:05.493 DEBG Using self rate limiting params         config: OutboundRateLimiterConfig { ping: 2/10s, metadata: 1/15s, status: 5/15s, goodbye: 1/10s, blocks_by_range: 1024/10s, blocks_by_root: 128/10s }, service: libp2p_rpc, service: libp2p
```

### CODE REVIEW

File: beacon_node/lighthouse_network/src/rpc/mod.rs [hunk 1]
```diff
@@ -277,6 +378,10 @@ where
         if !self.events.is_empty() {
             return Poll::Ready(self.events.remove(0));
         }
+        if let Poll::Ready(Some(Ok(expired))) = self.next_peer_request.poll_expired(cx) {
+            let (peer_id, protocol) = expired.into_inner();
+            self.send_rate_limited_request(peer_id, protocol)
```
Comment (AgeManning) [inline beacon_node/lighthouse_network/src/rpc/mod.rs]: I think we might run into issues where, the `send_rate_limited_request()` adds events to the `self.events` but the task then returns `Poll::Pending` and we won't actually send these events when they are ready until the task wakes up again. 

We should either move this code above the check for `self.events` or check that its not empty again, I think.

**Review by AgeManning — COMMENTED** [review]
This looks pretty good to me. 

let me know when you think its ready for a final review and possibility of merge.

**Comment by divagant-martian** [comment]
oh yeah, this was a prototype but I think I'll push it to the whole thing, so we are missing at least metrics and a flag I'll mark as wip for now

File: beacon_node/lighthouse_network/src/rpc/config.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+impl OutboundRateLimiterConfig {
+    pub const DEFAULT_PING_QUOTA: Quota = Quota::n_every(2, 10);
+    pub const DEFAULT_META_DATA_QUOTA: Quota = Quota::n_every(2, 5);
+    pub const DEFAULT_STATUS_QUOTA: Quota = Quota::n_every(5, 15);
+    pub const DEFAULT_GOODBYE_QUOTA: Quota = Quota::one_every(10);
+    pub const DEFAULT_BLOCKS_BY_RANGE_QUOTA: Quota =
+        Quota::n_every(methods::MAX_REQUEST_BLOCKS, 10);
+    pub const DEFAULT_BLOCKS_BY_ROOT_QUOTA: Quota = Quota::n_every(128, 10);
```
Comment (AgeManning) [inline beacon_node/lighthouse_network/src/rpc/config.rs:79]: Do you think we should use these defaults for the inbound limiter also? 

I would expect by default the outbound should match the inbound, right?

**Review by AgeManning — COMMENTED** [review]
Looks reasonable to me. :)

**Review by AgeManning — APPROVED** [review]

File: beacon_node/lighthouse_network/src/rpc/config.rs (hunk 2 above)
Comment (divagant-martian) [reply beacon_node/lighthouse_network/src/rpc/config.rs:79]: I'll update this when we add the option to turn off inbound rate limiting

**Pull request closed without merging**